### PR TITLE
Adding a cluster method - RO-4226

### PR DESCRIPTION
**Pull request by josemduarte**
This goes together with https://github.com/rcsb/py-rcsb_db/pull/69

@piehld any feedback appreciated, especially on style or other conventions to follow.

### CODE REVIEW

File: rcsb/utils/tests-seqalign/testMMseqsUtils.py [hunk 1]
```diff
@@ -255,6 +255,30 @@ def testAlignedRegions(self):
...
+        try:
+            inFile = os.path.join(self.__dataPath, "sabdab-targets.fa")
+            clustersOutFile = "/tmp/seqclusters-test.txt"
+            mmS = MMseqsUtils(cachePath=self.__workPath)
+            mmS.cluster(inFile, 50, 0.8, "easy-cluster", "/tmp", clustersOutFile)
+            self.assertTrue(os.path.exists(clustersOutFile))
+            # compare number of sequences in input and output
+            countSeqsIn = 0
+            countSeqsOut = 0
+            with open(inFile, "r") as f:
```
Comment (piehld) [inline rcsb/utils/tests-seqalign/testMMseqsUtils.py]:
```suggestion
            with open(inFile, "r", encoding="utf-8") as f:
```

File: rcsb/utils/tests-seqalign/testMMseqsUtils.py [hunk 2]
```diff
@@ -255,6 +255,30 @@ def testAlignedRegions(self):
...
+                for line in f.readlines():
+                    if line.startswith(">"):
+                        countSeqsIn = countSeqsIn + 1
+            with open(clustersOutFile, "r") as f:
```
Comment (piehld) [inline rcsb/utils/tests-seqalign/testMMseqsUtils.py]:
```suggestion
            with open(clustersOutFile, "r", encoding="utf-8") as f:
```

File: rcsb/utils/seqalign/MMseqsUtils.py [hunk 3]
```diff
@@ -718,3 +719,106 @@ def getAlignedRegions(self, cigarString, queryStart, targetStart):
...
+            entityList.append([len(cluster), sorted(cluster)])
+
+        self.__writeClusters(entityList, clusters_out_file)
+
+    @staticmethod
+    def __readClusters(tsvFile):
+        """
+           Read mmseqs2 cluster tsv output file
+        """
+        f = open(tsvFile, 'r')
```
Comment (piehld) [inline rcsb/utils/seqalign/MMseqsUtils.py]:
```suggestion
        f = open(tsvFile, "r", encoding="utf-8")
```

File: rcsb/utils/seqalign/MMseqsUtils.py [hunk 4]
```diff
@@ -718,3 +719,106 @@ def getAlignedRegions(self, cigarString, queryStart, targetStart):
...
+                clusterMap[cluster[0]].append(cluster[1])
+            else:
+                clusterMap[cluster[0]] = [cluster[1]]
+
+        return list(clusterMap.values())
+
+    @staticmethod
+    def __writeClusters(clusterList, clustersOutFile):
+        sortList = sorted(clusterList, key=lambda data: (-data[0], data[1]))
+        f = open(clustersOutFile, 'w')
```
Comment (piehld) [inline rcsb/utils/seqalign/MMseqsUtils.py]:
```suggestion
        f = open(clustersOutFile, "w", encoding="utf-8")
```

File: rcsb/utils/seqalign/MMseqsUtils.py [hunk 5]
```diff
@@ -718,3 +719,106 @@ def getAlignedRegions(self, cigarString, queryStart, targetStart):
...
+            "--cluster-mode", "1",
+            "--threads", str(workers)
+        ]
+        if strategy == "easy-cluster":
+            cmdArgs.extend([
+                "-s", "8",
+                "--max-seqs", "1000"
+            ])
+
+        logger.info("Will run command '%s %s'" % (self.__mmseqs2BinPath, " ".join(cmdArgs)))
```
Comment (piehld) [inline rcsb/utils/seqalign/MMseqsUtils.py]:
```suggestion
        logger.info("Will run command '%s %s'", self.__mmseqs2BinPath, " ".join(cmdArgs))
```

**Review by piehld — CHANGES_REQUESTED** [review]
Thanks @josemduarte, not sure if you're done with your changes or not, but I made some suggestions to fix the linting issues you're still having. Also, one other formatting guideline the ExDB code generally follows is to use double quotes instead of single quotes (when possible). I know it's kind of trivial, but but it's consistent enough that the single quotes really stand out to me...

Once you have the code ready for my review again, let me know and I'll take a deeper look into it.

**Comment by josemduarte** [comment]
Thanks for the reviewing and for the suggestions @piehld 

As explained in ticket I've decided to port to java, so that we can reuse the infrastructure that we had already. See comments in RO-4226

Closing in favor of https://github.com/rcsb/rcsb-redwood/pull/131

**Pull request closed without merging**